atc/db/pipeline_lifecycle.go

```diff
@@ -0,0 +1,87 @@
+package db
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+type PipelineLifecycle interface {
+	ArchiveAbandonedPipelines() (int, error)
+}
+
+func NewPipelineLifecycle(conn Conn, lockFactory lock.LockFactory) PipelineLifecycle {
+	return &pipelineLifecycle{
+		conn:        conn,
+		lockFactory: lockFactory,
+	}
+}
+
+type pipelineLifecycle struct {
+	conn        Conn
+	lockFactory lock.LockFactory
+}
+
+func (pl *pipelineLifecycle) ArchiveAbandonedPipelines() (int, error) {
```

Comment:
One alternative to having a GC component would be to perform the query whenever a build finishes successfully. That way, we wouldn't need to perform the `successful_builds` subquery in the first place - we could just use the job ID from the finished build.

Did you consider about going with that approach? One benefit is that the pipeline archiving would be immediate. Not sure what the performance impacts would be on an environment with lots of builds running constantly, though I imagine the query would be pretty cheap if just on a single `job_id`

Reply:
Didn't think of doing this approach, though I'm hesitant to add a query that would be run at the end of **every** build. I think this would be problematic for people running large clusters where there are hundreds of builds finishing every minute. I feel like the query would be a no-op for ~99% of the builds too.

You can see our thought process here for some other alternatives: https://github.com/concourse/concourse/issues/5699#issuecomment-654475693

Reply:
> One alternative to having a GC component would be to perform the query whenever a build finishes successfully.

That would have a problem on small systems if you delete a pipeline (or remove a job) and there's no builds running at the time. We can split it into 3 smaller queries that get triggered when one of the 3 conditions are met, but that would spread the logic over into multiple places.

Reply:
True that the query would be a no-op the majority of the time, but I imagine with an index on `p.parent_job_id`, the query would be really cheap - the `WHERE p.parent_job_id = $1` would result in 0 rows and short-circuit for all the no-op jobs, and would result in only a few rows when it does run `set_pipeline`. So the query for a build scales with the number of pipelines that the job sets, which seems appropriate.

With the GC/interval approach, it seems like there's a constant cost every GC tick (with some scary terms in the `EXPLAIN ANALYZE` like "Nested Loop Left Join" and "Sort Method: external merge"). Maybe that's okay (I'm definitely not an expert on SQL optimization), but it does seem like a pretty expensive query to be running every 30 seconds.

Worth noting that even with running the query in `Build.Finish()`, we'd still want to have some sort of GC interval going on to detect the *other* ways pipelines can get archived (parent pipeline destroyed, job no longer active) - though these would likely be much cheaper, given that we wouldn't need the `successful_builds` subquery.

Maybe this is something we can take a look into optimizing if we run into performance woes with this approach

Reply:
Hm I'm not 100% with my opinion but I think I would slightly prefer having the garbage collector archive pipelines purely in the sense that I would prefer having a separate component be the one that archives a pipeline rather than having the finishing of a build be overloaded with more responsibilities. But I also understand that builds can be the "owners" of a pipeline so maybe it does make sense?

I do think there is a concern about the responsiveness of the archiving if it was done by gc, where the users would have to wait for up to 10 seconds to see their pipeline be archived after a build finishes. I'm thinking maybe we can have a notification that gets sent to the component that archives the pipelines? But there is a worry that if we do keep a interval of 10 seconds, because there are so many builds finishing in a large deployment having a notification sent will not do much because the gc component can only run once every 10 seconds anyways. Maybe is there a way to make it only send a notification if it is a build that owns a pipeline? And that way we can trust the notification that comes in and run it everytime?

Reply:
@clarafu the thing that strikes me as odd if we were to send a notification to GC when a build finishes (and why I prefer performing the query in `Finish`) is that we already *know* what (single) `job_id` we need to check when the build finishes - with the separate GC component, we'd just be choosing to ignore that information and instead consider all the jobs.

I guess we could send along the notification bus the `job_id` that needs to be checked - I've implemented this a few times in my experimental PRs: https://github.com/concourse/concourse/pull/5651/commits/85e285272b1f38bf243cb1ace1d2386fa5a1b7ec (the difficulty being we then have to queue up notifications so we don't lose information). But that seems more complicated than it should be.

Reply:
Ya I agree with you, I think its unnecessary. My concern is probably it feeling a bit messy having a few different places where a pipeline can be archived (api, build finishing and gc) and it might feel weird as a user if sometimes their pipeline gets archived immediately (build finishing) but others it takes up to 10 seconds to archive (job is set as inactive). As a thought experiment, would it be possible to always archive pipelines always immediately? This means we no longer need a gc for archiving pipelines and we will always just archive when necessary (build finishing, parent pipeline destroyed, etc).

As a side note: I think if a build finishes and a pipeline is archived as a result, I think we will probably want to also immediately archive any pipelines that were set by that archived pipeline.

Reply:
I have the same thoughts as Clara
> ...having a few different places where a pipeline can be archived (api, build finishing and gc)...

Also makes it harder for us to maintain and think about the "automatic archiving feature".

We're going to stick with the gc approach. I think users are use to Concourse taking some time to do things, like job scheduling, running checks, etc.
Waiting max 30s probably won't feel abnormal to them and by the time they ask "why hasn't my pipeline been archived yet?" it'll probably be gone.

Reply:
My concerns are now less about the immediacy of archiving and more about choosing to ignore info that would make the query much cheaper (scoped to a single job rather than all jobs). If we make the query cheap across all jobs, though, then the extra complexity probably won't be worth it

Reply:
I think my only concern about the difference between Concourse taking some time to do things like job scheduling or running checks and archiving pipelines in gc is the fact that we show the users that something is happening in the background (build is pending or check is running). With archiving pipelines, the user won't see anything in the UI that shows that the archive is pending and it might just seem laggy that the archiving takes long to happen or maybe even broken because there is no indication that it is waiting to be archived. What do you think @taylorsilva? But maybe I'm just worried about something that could be a UI concern of being more transparent of when it is waiting to be archived?

Reply:
I agree with the UI/UX concerns here. Feels like automatic archiving is not very transparent about how it works or what it's doing. I'll be fine with that though if we detail how it works and when it GC's in the docs so users can find an answer or we can at least point to an answer when users ask.

Overall I still prefer the GC approach for its simplicity. I think we can still make a performant query with the GC approach. If we need to improve the performance of the query there are options. Right now we know the CTE part of the query is slow. The newer query (subquery with `LIMIT 1`) looks like it'll be faster but we need to test it which is what we plan to do today. We can also still go the denormalization route if we need even better performance (at the cost of data integrity and duplication). I'd prefer trying those routes first before splitting up each case where we should automatically archive into various parts of the code base.

Reply:
if theres a delay in the UI, could I just kick off another build without knowing a piplines going to be archived the next min? (Might be a dumb question, Idk what archivable piplines looks like and if they let you kick off more than 1 build)

Reply:
@kcmannem yeah good point. I guess we could add a check in the “reject archived” wrappa that would determine if the pipeline would be archived in the next GC tick, and that would prevent performing any operations that are disallowed on archived pipelines (e.g triggering builds)

I hadn’t really thought about what to do about pipelines that have active builds when we try to archive them...maybe the GC could mark running pipelines as “awaiting archive”, and they could get swept in the next GC tick? We’d prevent scheduling on pipelines that are awaiting archived

Reply:
Or could we just archive the pipeline right away? I guess it may not be a problem archiving a pipeline with a running build, since the build could still complete

Reply:
Yeah, it should be OK to archive while a build is running; the build will eventually complete and the user will still be able to see the build if they navigate to the archived pipeline. It should also be safe to clear out the config, since by that point the build is independent of it and just running a build plan.

I do think that having the build complete and then the pipelines remain active and become archived some amount of time later doesn't feel great, which has already been pointed out. 🤔 The user may think Concourse decided not to archive it and wonder why. Seems like a potential source of repeated minor anxiety/uncertainty.

Archiving immediately seems a lot better since it removes that window of time, but TBH still not *super* great because it still may not be obvious that the build caused it to happen.

What I'm about to suggest is totally scope creep and belongs in a separate issue if we decide to look into it, but: if we could perform the archiving operations at the end of the build *and show it happening in the build plan*, that would be pretty awesome. It's a bit tricky because it's not just something you can do in the `set_pipeline` step - it has to be done after all `set_pipeline` steps have completed so we know the full set to diff. I'm tempted to try to frame this as a dynamic build plan, because such a mechanism could be helpful in other scenarios in the future (e.g. using the `across` step with a set of values that were generated within the build).

🤔 🤔 🤔

Reply:
Like this? Being appended to the end of a build plan? Seems nice that it's explicit.

```
on_complete:
  - archive_pipeline
```

Reply:
@vito rather than using dynamic build plan, could we just create a new build event that represents archiving a pipeline? So the archiving doesn’t have a build plan, but we can still be explicit about what happened when looking at the build in the UI/Fly?

Can see the value in supporting dynamic build plans for across, though

Reply:
We decided to add onto `build.Finish()` so pipelines that are no longer being set by the parent pipeline are removed immediately at the end of a build.

All other cases are covered by the GC approach.

Reply:
@kcmannem I was thinking that Concourse would automatically do it - by build plan I meant the internal `atc.Plan`, not the config `plan:`.

@aoldershaw Yeah that'd be a more lightweight option!

@taylorsilva Sounds good, let's look for feedback from users about this before committing to UI changes - at least it happens immediately!